File: POS software/cpp/src/pages/settingspage.cpp
#include "settingspage.h"
#include "../database.h"
#include "../styles.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QPushButton>
#include <QFrame>
#include <QMessageBox>
#include <QScrollArea>
#include <QFile>
#include <QCoreApplication>
#include <QDir>

SettingsPage::SettingsPage(QWidget* parent) : QWidget(parent) {
    buildUI();
    loadSettings();
}

void SettingsPage::buildUI() {
    auto* scroll = new QScrollArea(this);
    scroll->setWidgetResizable(true);
    scroll->setFrameShape(QFrame::NoFrame);
    scroll->setStyleSheet("background: transparent;");

    auto* content = new QWidget;
    auto* lay = new QVBoxLayout(content);
    lay->setContentsMargins(28, 24, 28, 24);
    lay->setSpacing(20);

    auto* title = new QLabel("\xE2\x9A\x99\xEF\xB8\x8F  Pengaturan");
    title->setStyleSheet(QStringLiteral(
        "font-size: 22px; font-weight: 700; color: %1;").arg(Style::TEXT_PRIMARY));
    lay->addWidget(title);

    // ── Store info card ──────────────────────────────────────
    auto* storeCard = new QFrame;
    storeCard->setStyleSheet(QStringLiteral(
        "QFrame{background-color:%1;border-radius:16px;}").arg(Style::BG_CARD));
    storeCard->setMaximumWidth(600);
    auto* storeForm = new QFormLayout(storeCard);
    storeForm->setContentsMargins(24, 20, 24, 20);
    storeForm->setSpacing(12);

    auto* storeTitle = new QLabel("Informasi Toko");
    storeTitle->setStyleSheet(QStringLiteral(
        "font-size: 16px; font-weight: 700; color: %1;").arg(Style::TEXT_PRIMARY));
    storeForm->addRow(storeTitle);

    m_storeName = new QLineEdit;    m_storeName->setMinimumHeight(36);
    m_storeAddress = new QLineEdit; m_storeAddress->setMinimumHeight(36);
    m_storePhone = new QLineEdit;   m_storePhone->setMinimumHeight(36);
    storeForm->addRow("Nama Toko:", m_storeName);
    storeForm->addRow("Alamat:", m_storeAddress);
    storeForm->addRow("Telepon:", m_storePhone);
    lay->addWidget(storeCard);

    // ── Tax & receipt card ───────────────────────────────────
    auto* taxCard = new QFrame;
    taxCard->setStyleSheet(QStringLiteral(
        "QFrame{background-color:%1;border-radius:16px;}").arg(Style::BG_CARD));
    taxCard->setMaximumWidth(600);
    auto* taxForm = new QFormLayout(taxCard);
    taxForm->setContentsMargins(24, 20, 24, 20);
    taxForm->setSpacing(12);

    auto* taxTitle = new QLabel("Pajak & Struk");
    taxTitle->setStyleSheet(QStringLiteral(
        "font-size: 16px; font-weight: 700; color: %1;").arg(Style::TEXT_PRIMARY));
    taxForm->addRow(taxTitle);

    m_taxRate = new QLineEdit;        m_taxRate->setMinimumHeight(36);
    m_receiptFooter = new QLineEdit;  m_receiptFooter->setMinimumHeight(36);
    m_lowStockThreshold = new QLineEdit; m_lowStockThreshold->setMinimumHeight(36);
    taxForm->addRow("Pajak (%):", m_taxRate);
    taxForm->addRow("Footer Struk:", m_receiptFooter);
    taxForm->addRow("Threshold Stok Rendah:", m_lowStockThreshold);
    lay->addWidget(taxCard);

    // ── Save button ──────────────────────────────────────────
    auto* saveBtn = new QPushButton("Simpan Pengaturan");
    saveBtn->setMinimumHeight(42);
    saveBtn->setMaximumWidth(600);
    saveBtn->setCursor(Qt::PointingHandCursor);
    connect(saveBtn, &QPushButton::clicked, this, &SettingsPage::saveSettings);
    lay->addWidget(saveBtn);

    // ── App info card ────────────────────────────────────────
    auto* infoCard = new QFrame;
    infoCard->setStyleSheet(QStringLiteral(
        "QFrame{background-color:%1;border-radius:16px;}").arg(Style::BG_CARD));
    infoCard->setMaximumWidth(600);
    auto* infoLay = new QVBoxLayout(infoCard);
    infoLay->setContentsMargins(24, 18, 24, 18);
    infoLay->setSpacing(8);
    auto* appTitle = new QLabel("Tentang Aplikasi");
    appTitle->setStyleSheet(QStringLiteral(
        "font-size: 16px; font-weight: 700; color: %1;").arg(Style::TEXT_PRIMARY));
    infoLay->addWidget(appTitle);

    auto addInfo = [&](const QString& label, const QString& value) {
        auto* row = new QHBoxLayout;
        auto* l = new QLabel(label);
        l->setStyleSheet(QStringLiteral("color:%1;font-size:12px;").arg(Style::TEXT_SECONDARY));
        auto* v = new QLabel(value);
        v->setStyleSheet(QStringLiteral("color:%1;font-size:12px;").arg(Style::TEXT_PRIMARY));
        row->addWidget(l);
        row->addWidget(v);
        row->addStretch();
        infoLay->addLayout(row);
    };
    addInfo("Aplikasi:", "POS System v2.0");
    addInfo("Framework:", "Qt 6 (C++)");
    addInfo("Database:", "SQLite");
    addInfo("Developer:", "POS Team");
    lay->addWidget(infoCard);

    // ── Danger zone ──────────────────────────────────────────
    auto* dangerCard = new QFrame;
    dangerCard->setStyleSheet(QStringLiteral(
        "QFrame{background-color:%1;border:1px solid %2;border-radius:16px;}")
        .arg(Style::BG_CARD, Style::DANGER));
    dangerCard->setMaximumWidth(600);
    auto* dangerLay = new QVBoxLayout(dangerCard);
    dangerLay->setContentsMargins(24, 18, 24, 18);
    dangerLay->setSpacing(10);
    auto* dangerTitle = new QLabel("\xE2\x9A\xA0\xEF\xB8\x8F  Zona Berbahaya");
    dangerTitle->setStyleSheet(QStringLiteral(
        "font-size: 16px; font-weight: 700; color: %1;").arg(Style::DANGER));
    dangerLay->addWidget(dangerTitle);
    auto* dangerDesc = new QLabel("Reset database akan menghapus semua data dan mengembalikan ke default.");
    dangerDesc->setStyleSheet(QStringLiteral("color:%1;font-size:12px;").arg(Style::TEXT_SECONDARY));
    dangerDesc->setWordWrap(true);
    dangerLay->addWidget(dangerDesc);
    auto* resetBtn = new QPushButton("Reset Database");
    resetBtn->setMinimumHeight(38);
    resetBtn->setStyleSheet(Style::dangerBtnStyle());
    resetBtn->setCursor(Qt::PointingHandCursor);
    connect(resetBtn, &QPushButton::clicked, this, &SettingsPage::resetDatabase);
    dangerLay->addWidget(resetBtn);
    lay->addWidget(dangerCard);

    lay->addStretch();

    scroll->setWidget(content);
    auto* rootLay = new QVBoxLayout(this);
    rootLay->setContentsMargins(0, 0, 0, 0);
    rootLay->addWidget(scroll);
}

void SettingsPage::loadSettings() {
    auto s = Database::instance().getSettings();
    m_storeName->setText(s["store_name"].toString());
    m_storeAddress->setText(s["store_address"].toString());
    m_storePhone->setText(s["store_phone"].toString());
    m_taxRate->setText(QString::number(s["tax_rate"].toDouble()));
    m_receiptFooter->setText(s["receipt_footer"].toString());
    m_lowStockThreshold->setText(QString::number(s["low_stock_threshold"].toInt()));
}

void SettingsPage::saveSettings() {
    QVariantMap s;
    s["store_name"]          = m_storeName->text().trimmed();
    s["store_address"]       = m_storeAddress->text().trimmed();
    s["store_phone"]         = m_storePhone->text().trimmed();
    s["tax_rate"]            = m_taxRate->text().toDouble();
    s["receipt_footer"]      = m_receiptFooter->text().trimmed();
    s["low_stock_threshold"] = m_lowStockThreshold->text().toInt();

    if (Database::instance().updateSettings(s))
        QMessageBox::information(this, "Berhasil", "Pengaturan berhasil disimpan!");
    else
        QMessageBox::critical(this, "Error", "Gagal menyimpan pengaturan!");
}

void SettingsPage::resetDatabase() {
    auto r = QMessageBox::warning(this, "Konfirmasi",
        "PERINGATAN: Semua data akan dihapus!\n\nYakin ingin mereset database?",
        QMessageBox::Yes | QMessageBox::No);
    if (r != QMessageBox::Yes) return;

    Database::instance().close();
    QString dbPath = QDir(QCoreApplication::applicationDirPath()).filePath("pos_database.db");
    QFile::remove(dbPath);
    Database::instance().initialize();
    loadSettings();
    QMessageBox::information(this, "Berhasil", "Database berhasil direset!");
}

File: POS software/pages/settings_page.py
"""
Settings Page - Store settings, tax, receipt, etc.
"""
import customtkinter as ctk
from tkinter import messagebox
import database as db

COLORS = {
    "bg_dark": "#0f0f14",
    "bg_card": "#1e1e2e",
    "bg_card_hover": "#262638",
    "bg_entry": "#252535",
    "accent": "#6c63ff",
    "accent_hover": "#7b73ff",
    "success": "#2ecc71",
    "success_hover": "#27ae60",
    "danger": "#e74c3c",
    "text_primary": "#e8e8f0",
    "text_secondary": "#a9a9c0",
    "text_muted": "#6c6c88",
    "border": "#2a2a3d",
    "header_bg": "#1a1a28",
}


class SettingsPage:
    def __init__(self, parent, app):
        self.parent = parent
        self.app = app
        self._build()

    def _build(self):
        header = ctk.CTkFrame(self.parent, fg_color=COLORS["header_bg"], height=64, corner_radius=0)
        header.pack(fill="x")
        header.pack_propagate(False)

        ctk.CTkLabel(header, text="⚙  Pengaturan", font=("Segoe UI Bold", 20),
                     text_color=COLORS["text_primary"]).pack(side="left", padx=24, pady=16)

        scroll = ctk.CTkScrollableFrame(self.parent, fg_color=COLORS["bg_dark"],
                                         scrollbar_button_color="#3a3a50",
                                         scrollbar_button_hover_color=COLORS["accent"])
        scroll.pack(fill="both", expand=True, padx=20, pady=12)

        settings = db.get_all_settings()
        self.entries = {}

        # ── Store Info ───────────────────────────────────────
        self._section_header(scroll, "🏪 Informasi Toko")
        store_card = ctk.CTkFrame(scroll, fg_color=COLORS["bg_card"], corner_radius=14)
        store_card.pack(fill="x", padx=4, pady=(0, 16))

        for key, label, default in [
            ("store_name", "Nama Toko", "TOKO SAYA"),
            ("store_address", "Alamat Toko", ""),
            ("store_phone", "Telepon Toko", ""),
        ]:
            self._add_setting_field(store_card, key, label, settings.get(key, default))

        # ── Tax & Payment ────────────────────────────────────
        self._section_header(scroll, "💰 Pajak & Pembayaran")
        tax_card = ctk.CTkFrame(scroll, fg_color=COLORS["bg_card"], corner_radius=14)
        tax_card.pack(fill="x", padx=4, pady=(0, 16))

        self._add_setting_field(tax_card, "tax_rate", "Pajak (%)", settings.get("tax_rate", "11"))
        self._add_setting_field(tax_card, "currency_symbol", "Simbol Mata Uang", settings.get("currency_symbol", "Rp"))

        # ── Receipt ──────────────────────────────────────────
        self._section_header(scroll, "🧾 Struk / Receipt")
        receipt_card = ctk.CTkFrame(scroll, fg_color=COLORS["bg_card"], corner_radius=14)
        receipt_card.pack(fill="x", padx=4, pady=(0, 16))

        self._add_setting_field(receipt_card, "receipt_footer", "Footer Struk",
                                settings.get("receipt_footer", "Terima kasih atas kunjungan Anda!"))

        # ── Inventory ────────────────────────────────────────
        self._section_header(scroll, "📦 Inventaris")
        inv_card = ctk.CTkFrame(scroll, fg_color=COLORS["bg_card"], corner_radius=14)
        inv_card.pack(fill="x", padx=4, pady=(0, 16))

        self._add_setting_field(inv_card, "low_stock_alert", "Batas Peringatan Stok Rendah",
                                settings.get("low_stock_alert", "5"))

        # Save button
        ctk.CTkButton(scroll, text="💾  Simpan Pengaturan", height=48, corner_radius=12,
                      fg_color=COLORS["accent"], hover_color=COLORS["accent_hover"],
                      font=("Segoe UI Bold", 16), command=self._save).pack(fill="x", padx=4, pady=(8, 16))

        # ── App Info ─────────────────────────────────────────
        info_card = ctk.CTkFrame(scroll, fg_color=COLORS["bg_card"], corner_radius=14)
        info_card.pack(fill="x", padx=4, pady=(0, 16))

        ctk.CTkLabel(info_card, text="ℹ️ Tentang Aplikasi", font=("Segoe UI Semibold", 14),
                     text_color=COLORS["text_primary"]).pack(padx=20, pady=(16, 8), anchor="w")

        info_items = [
            ("Nama Aplikasi", "POS System - Point of Sale"),
            ("Versi", "1.0.0"),
            ("Framework", "Python + CustomTkinter"),
            ("Database", "SQLite3"),
            ("Dibuat dengan", "❤️ Python"),
        ]
        for label, val in info_items:
            row = ctk.CTkFrame(info_card, fg_color="transparent", height=28)
            row.pack(fill="x", padx=20, pady=1)
            row.pack_propagate(False)
            ctk.CTkLabel(row, text=label, font=("Segoe UI", 11),
                         text_color=COLORS["text_secondary"]).pack(side="left")
            ctk.CTkLabel(row, text=val, font=("Segoe UI", 11),
                         text_color=COLORS["text_primary"]).pack(side="right")

        ctk.CTkFrame(info_card, fg_color="transparent", height=12).pack()

        # Reset button
        ctk.CTkButton(scroll, text="🔄  Reset Database (Hati-hati!)", height=40, corner_radius=10,
                      fg_color=COLORS["danger"], hover_color="#c0392b",
                      font=("Segoe UI Semibold", 13),
                      command=self._reset_db).pack(fill="x", padx=4, pady=(0, 24))

    def _section_header(self, parent, text):
        ctk.CTkLabel(parent, text=text, font=("Segoe UI Bold", 15),
                     text_color=COLORS["text_primary"]).pack(anchor="w", padx=8, pady=(12, 6))

    def _add_setting_field(self, parent, key, label, value):
        frame = ctk.CTkFrame(parent, fg_color="transparent")
        frame.pack(fill="x", padx=16, pady=6)

        ctk.CTkLabel(frame, text=label, font=("Segoe UI", 12),
                     text_color=COLORS["text_secondary"]).pack(anchor="w", pady=(0, 2))
        entry = ctk.CTkEntry(frame, height=38, corner_radius=8,
                             fg_color=COLORS["bg_entry"], border_color=COLORS["border"],
                             text_color=COLORS["text_primary"], font=("Segoe UI", 12))
        entry.pack(fill="x")
        entry.insert(0, str(value))
        self.entries[key] = entry

    def _save(self):
        for key, entry in self.entries.items():
            db.update_setting(key, entry.get().strip())
        messagebox.showinfo("Sukses", "Pengaturan berhasil disimpan!\nBeberapa perubahan memerlukan restart aplikasi.")

    def _reset_db(self):
        if messagebox.askyesno("Reset Database",
                               "PERINGATAN: Ini akan menghapus SEMUA data!\n\nApakah Anda yakin?",
                               icon="warning"):
            if messagebox.askyesno("Konfirmasi Ulang",
                                   "Tindakan ini tidak dapat dibatalkan.\nKonfirmasi reset database?",
                                   icon="warning"):
                import os
                db_path = db.DB_PATH
                if os.path.exists(db_path):
                    conn = db.get_connection()
                    conn.close()
                    os.remove(db_path)
                db.init_database()
                messagebox.showinfo("Sukses", "Database berhasil di-reset!\nSilakan restart aplikasi.")
